bin/tx-prover/src/proxy/mod.rs

```diff
@@ -74,77 +74,65 @@ impl LoadBalancer {
+static QUEUE: Lazy<RwLock<Vec<u64>>> = Lazy::new(|| RwLock::new(Vec::new()));
 
 /// Custom context for the request/response lifecycle
-/// We use this context to keep track of the number of tries for a request.
-pub struct TriesCounter {
+/// We use this context to keep track of the number of tries for a request, the unique ID for the
+/// request, and the worker that will process the request.
+pub struct RequestContext {
+    /// Number of tries for the request
     tries: usize,
+    /// Unique ID for the request
+    request_id: u64,
+    /// Worker that will process the request
+    worker: Option<Backend>,
 }
 
 /// Implements load-balancing of incoming requests across a pool of workers.
 ///
 /// At the backend-level, a request lifecycle works as follows:
 /// - When a new requests arrives, [LoadBalancer::request_filter()] method is called. In this method
-///   we apply IP-based rate-limiting to the request and assign a unique ID to it.
+///   we apply IP-based rate-limiting to the request and check if the queue is full.
 /// - Next, the [Self::upstream_peer()] method is called. We use it to figure out which worker will
-///   process the request. Inside `upstream_peer()`, we pick a worker in a round-robin fashion and
-///   add the request to the queue of requests for that worker. Once the request gets to the front
-///   of the queue, we forward it to the worker. This step is also in charge of assinging the
-///   timeouts and enabling HTTP/2. Finally, we establish a connection with the worker.
+///   process the request. Inside `upstream_peer()`, we add the request to the queue of requests.
+///  Once the request gets to the front of the queue, we forward it to an available worker. This
+///  step is also in charge of setting the SNI, timeouts, and enabling HTTP/2. Finally, we
+///  establish a connection with the worker.
 /// - Before sending the request to the upstream server and if the connection succeed, the
 ///   [Self::upstream_request_filter()] method is called. In this method, we ensure that the correct
 ///   headers are forwarded for gRPC requests.
 /// - If the connection fails, the [Self::fail_to_connect()] method is called. In this method, we
 ///   retry the request [self.max_retries_per_request] times.
 /// - Once the worker processes the request (either successfully or with a failure),
-///   [Self::logging()] method is called. In this method, we remove the request from the worker's
-///   queue, allowing the worker to process the next request.
+///   [Self::logging()] method is called. In this method, we log the request lifecycle and set the
+///   worker as available.
 #[async_trait]
 impl ProxyHttp for LoadBalancer {
-    type CTX = TriesCounter;
+    type CTX = RequestContext;
     fn new_ctx(&self) -> Self::CTX {
-        TriesCounter { tries: 0 }
+        RequestContext {
+            tries: 0,
+            request_id: rand::random::<u64>(),
+            worker: None,
+        }
```

Comment:
nit: Might be worth adding a `RequestContext::new()` that creates and initializes the request